[ONNX] Support ONNXFakeContext with op_level_debug

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105874

Previous to the PR, op_level_debug doesn't support OnnxFakeConext because it relies on real tensor in args to do shape type inference propagation in fx graph to get static shapes helping simulating the op args/kwargs. However, OnnxFakeContext will fakify the args/kwargs at the very begining, so the op_level_debug can't have the static_shapes to utilize.

This PR uses SymInt API: `has_hint` and `hint_int` to fully replace the functionality of shape type inference propagation. The static shapes are obtained through SymInt. Therefore, the pass `ShapeInferenceWithFakeTensor` is eliminated.

Also moved the args/kwargs processing into op_validation to live under the rule `op_level_debug`.

titaiwangms committed:
[ONNX] Support ONNXFakeContext with op_level_debug

[ghstack-poisoned]

thiagocrepaldi commented on torch/onnx/_internal/exporter.py:
```diff
@@ -890,10 +890,10 @@ def pre_export_passes(
-        module = passes.ShapeInferenceWithFakeTensor(diagnostic_context, module).run(
-            *fx_module_args
-        )
+    # if options.op_level_debug:
+    #     module = passes.ShapeInferenceWithFakeTensor(diagnostic_context, module).run(
+    #         *fx_module_args
+    #     )
```
```suggestion
```

titaiwangms commented on torch/onnx/_internal/exporter.py:
```diff
@@ -890,10 +890,10 @@ def pre_export_passes(
+    # if options.op_level_debug:
+    #     module = passes.ShapeInferenceWithFakeTensor(diagnostic_context, module).run(
+    #         *fx_module_args
+    #     )
```
Delete

BowenBao replied:
nit: reminder to delete this + comment + the pass code itself.

thiagocrepaldi commented on torch/onnx/_internal/fx/fx_onnx_interpreter.py:
```diff
@@ -405,6 +405,7 @@ def run_node(
                 fx_name_to_onnxscript_value,
                 onnxfunction_dispatcher,
                 op_level_debug,
+                fx_graph_module,
```
nit: in other files, we have been using the `fx_module` to refer to the fx graph module and `torch_module` when we refer to the `torch.nn.Module` module

titaiwangms replied:
I am following the variable name from `fx_onnx_interpreter.py`. I think it's fine.

thiagocrepaldi commented on torch/onnx/_internal/fx/op_validation.py:
```diff
@@ -197,38 +234,36 @@ def generate_random_tensors(shape: torch.Size, dtype: torch.dtype):
-                    f"arg.meta['static_shape']: {fake_tensor}; type(arg.meta['static_shape']): "
-                    f"{type(fake_tensor)}. Sym type is not supported in op_level_debug."
-                )
+            elif symbolic_shapes.has_hint(fake_tensor):
```
good to see that approach worked :)

thiagocrepaldi commented on torch/onnx/_internal/fx/op_validation.py:
```diff
@@ -165,8 +182,28 @@ def validate_op_between_ort_torch(
 
 
+@_beartype.beartype
+def _convert_symint_to_int_in_shape(shape: torch.Size) -> torch.Size:
```
```suggestion
def _try_convert_symint_to_int_in_shape(shape: torch.Size) -> torch.Size:
```
This may or may not convert a symint, so a `try` might be more accurate? not a strong preference though

BowenBao replied:
it raises when it cannot, so maybe add `raises` into docstring and keep the name? Usually we name with `_try` or `_maybe` if it gives up without raising.

titaiwangms replied:
Keep the name and added docstring

thiagocrepaldi reviewed (approved):
LGTM, just minor nits

BowenBao commented on torch/onnx/_internal/fx/op_validation.py:
```diff
@@ -197,38 +234,36 @@ def generate_random_tensors(shape: torch.Size, dtype: torch.dtype):
-    for arg in complete_args:
+    for arg in fx_args:
         if isinstance(arg, torch.fx.Node):
+            fake_tensor = arg.meta.get("val", None)
```
```suggestion
            fake_tensor = arg.meta.get("val")
```
nit: equivalent

justinchuby commented on torch/onnx/_internal/fx/op_validation.py:
```diff
@@ -165,8 +182,28 @@ def validate_op_between_ort_torch(
+                list_int_shape.append(symbolic_shapes.hint_int(dim))
+            else:
+                raise ValueError(
+                    f"Unexpected SymInt found in shape. SymInt: {dim}; "
```
Is this unexpected, or is it a case we can expect but cannot handle? Just to make sure for clarity.

BowenBao replied:
nice catch, overlooked this one. Please reveal keyword "unbacked", will easily lead to many helpful resources such as this https://pytorch-dev-podcast.simplecast.com/episodes/unbacked-symints-oyqu0_P6

titaiwangms replied:
Done

titaiwangms committed:
Update on "[ONNX] Support ONNXFakeContext with op_level_debug"


Previous to the PR, op_level_debug doesn't support OnnxFakeConext because it relies on real tensor in args to do shape type inference propagation in fx graph to get static shapes helping simulating the op args/kwargs. However, OnnxFakeContext will fakify the args/kwargs at the very begining, so the op_level_debug can't have the static_shapes to utilize.

This PR uses SymInt API: `has_hint` and `hint_int` to fully replace the functionality of shape type inference propagation. The static shapes are obtained through SymInt. Therefore, the pass `ShapeInferenceWithFakeTensor` is eliminated.

Also moved the args/kwargs processing into op_validation to live under the rule `op_level_debug`.

[ghstack-poisoned]
